pkg/comp-functions/functions/spks/spksredis/tls.go

```diff
@@ -0,0 +1,210 @@
+	}
+
+	svc.AddObservedConnectionDetails(comp.GetName() + "-server-cert")
+
+	return nil
+}
+
+func scaleRedisRelease(svc *runtime.ServiceRuntime, comp *spksv1alpha1.CompositeRedisInstance) error {
+	obsRelease := &xhelm.Release{}
+
+	err := svc.GetObservedComposedResource(obsRelease, "redis-chart")
+	if err != nil {
+		return fmt.Errorf("cannot get observed redis release: %w", err)
+	}
+
+	desRelease := &xhelm.Release{}
+
+	err = svc.GetDesiredComposedResourceByName(desRelease, "redis-chart")
+	if err != nil {
+		return fmt.Errorf("cannot get desired redis release: %w", err)
+	}
+
+	if getRedisReleaseTLSStatus(obsRelease, svc) != getRedisReleaseTLSStatus(desRelease, svc) || isSTSScaling(obsRelease, comp, svc) {
+		err := addSTSObserver(comp, svc)
+		if err != nil {
+			return err
+		}
+
+		values, err := convertValuesToMap(desRelease.Spec.ForProvider.Values.Raw)
+		if err != nil {
+			return err
+		}
+
+		err = common.SetNestedObjectValue(values, []string{"replica", "replicaCount"}, 0)
+		if err != nil {
+			return err
+		}
+
+		rawValues, err := json.Marshal(values)
+		if err != nil {
+			return err
+		}
+
+		desRelease.Spec.ForProvider.Values.Raw = rawValues
+
+		err = updateScaleTimestamp(svc, comp)
+		if err != nil {
+			return fmt.Errorf("cannot update composite status: %w", err)
+		}
+
+	}
+
+	err = svc.SetDesiredComposedResourceWithName(desRelease, "redis-chart")
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func getRedisReleaseTLSStatus(release *xhelm.Release, svc *runtime.ServiceRuntime) bool {
+
+	rawValues := release.Spec.ForProvider.Values
+
+	values, err := convertValuesToMap(rawValues.Raw)
+	if err != nil {
+		svc.Log.Info("cannot marshal release tls status: " + err.Error())
+		return false
+	}
+
+	enabled, _, err := unstructured.NestedBool(values, "tls", "enabled")
+	if err != nil {
+		svc.Log.Info("cannot get tls status from values: " + err.Error())
+		return false
+	}
+
+	return enabled
+}
+
+func convertValuesToMap(rawValues []byte) (map[string]any, error) {
+	values := map[string]any{}
+	err := json.Unmarshal(rawValues, &values)
+	if err != nil {
+		return nil, err
+	}
+
+	return values, nil
+}
+
+func addSTSObserver(comp *spksv1alpha1.CompositeRedisInstance, svc *runtime.ServiceRuntime) error {
+	sts := &appsv1.StatefulSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "redis-node",
+			Namespace: comp.GetName(),
+		},
+	}
+
+	err := svc.SetDesiredKubeObject(sts, comp.GetName()+"-sts-observer", runtime.KubeOptionObserve)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// isSTSScaling will return true, if the release it set to 0 replicas and
+// the sts hasn't fully scaled down yet.
+// This is necessary to ensure that we don't have mixed tls/non-tls members in the cluster
+// at the same time.
+func isSTSScaling(obsRelease *xhelm.Release, comp *spksv1alpha1.CompositeRedisInstance, svc *runtime.ServiceRuntime) bool {
```

Comment:
I wonder what is the consequence of leaving helm to normally rollout the changes? What happens when there are mixed pods with tls/non-tls?

Reply:
If you leave it to helm, or rather the sts controller, it will break the cluster and get stuck.

The STS controller will start with the highest numbered pod and restart it with the new settings. The pod will then attempt to connect to the other instances via TLS, which will fail. After some time the pod will force switch to be a master and at that point the Redis cluster won't be consistent anymore.

That's why we scale it manually first.

Reply:
perfect, thank you